### PR TITLE
[SPARK-5186] [MLLIB]  Vector.equals and Vector.hashCode are very inefficient

### PR DESCRIPTION
**Pull request by hhbyyh**
JIRA Issue: https://issues.apache.org/jira/browse/SPARK-5186

Currently SparseVector is using the inherited equals from Vector, which will create a full-size array for even the sparse vector. The pull request contains a specialized equals optimization that improves on both time and space. 
1. The implementation will be consistent with the original. Especially it will keep equality comparison between SparseVector and DenseVector.

### CODE REVIEW

**Comment by mengxr** [comment]
add to whitelist

**Comment by mengxr** [comment]
ok to test

**Comment by hhbyyh** [comment]
Thanks, can someone help review please?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 1]
```diff
@@ -449,6 +449,16 @@ class SparseVector(
   override def toString: String =
     "(%s,%s,%s)".format(size, indices.mkString("[", ",", "]"), values.mkString("[", ",", "]"))
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: You must override `hashCode` too!

Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Thanks @srowen for the comment. Glad to discuss it with someone.

_Vector: override def hashCode(): Int = util.Arrays.hashCode(this.toArray)_
 
I understand it's the general guideline to override `hashCode` at the same time.
Yet intentionally or not, the original code promises that `DenseVector` and `SparseVector` would return the same results of `equals` and `hashCode` for the same array content. And that makes some senses.

As in the description of the PR, I don’t want to introduce breaking changes. And if we want to keep the original design, the current implementation of `hashCode` in `Vector` is one of the best choices. That’s why `hashCode` was intentionally left out of the PR. (maybe I should add some comment)

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Yeah I was thinking of performance too. That is a fair point of course about not changing the semantics. I think the `hashCode` impl could be changed in both places to produce the same result while being faster for sparse vectors: what if the hash were defined only over nonzero entries? I think `equals` could likewise be sped up also for the case of comparing sparse to dense? that is I think this can be taken a step further than just optimizing sparse == sparse.

Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: +1 on @srowen 's suggestion. We can design a hashing scheme based on the size and nonzero entries (with their indices). But I think we could do that in a separate PR. Does it sound good?

Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Yes, the nonzero entries idea did cross my mind. Maybe it's overcautious that I think it might become a complexity if we want to have another kind of Vector in the future, which don't have handy internal structure to scan for the nonzero entries. Again, this can be overcautious. 

And the dense == sparse idea looks good, maybe that suits into a util method better as it would not introduce the existence of `DenseVector` to `SparseVector` and vice versa.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 2]
```diff
@@ -449,6 +449,16 @@ class SparseVector(
...
 
+  override def equals(other: Any): Boolean = {
+    other match {
+      case v: SparseVector =>
+        this.size == v.size &&
+          util.Arrays.equals(this.indices, v.indices) &&
+          util.Arrays.equals(this.values, v.values)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Just realized that this did change the behavior of the original `equals`, because we allow `values` to have explicit zeros. Maybe we can use `Vectors.sqdist(this, other) == 0` to implement `equals()`.

Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Nice suggestion. This should be your contribution and I'll close the PR. Thanks.

**Comment by hhbyyh** [comment]
There's a better idea from mengxr.

**Comment by srowen** [comment]
@hhbyyh Although I don't wish to speak for @mengxr I strongly suspect he would welcome you implementing the idea. I'm not sure we should use `sqdist` since it looks like it doesn't require vectors of the same size, and that seems essential for `equals` and `hashCode`? But I think it can indeed handle the cases of sparse vs dense similarly and efficiently.

**Comment by mengxr** [comment]
Agree with @srowen ! @hhbyyh It would be great if you want to re-open this PR and implement a faster and correct `Vector.equals`. If `sqdist` doesn't check vector sizes, we should add checks there. It is also nice to add a unit test in `VectorsSuite` to show it works for two sparse vectors with different `values` array but equal.

**Comment by hhbyyh** [comment]
Thanks, I'll try on it.

**Comment by hhbyyh** [comment]
Just send an update. I didn't use`sqdist` due to the performance concern. Since the original equals is actually a fail-fast comparison, yet `sqdist` will inevitably compute through the vectors even if the first element is different. The performance will be hard to accept for scenarios like doc2Vec over a large vocabulary. 

Current implementation is still based on the comparison for indices and values, just with the handling of the explicit 0. I gave some tests to the implementation and add a few ut. Any comment will be welcome!

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 3]
```diff
@@ -449,6 +449,31 @@ class SparseVector(
...
+          if (k1 == this.values.size || k2 == v.values.size) {
+            return (k1 == this.values.size && k2 == v.values.size) // check end alignment
+          }
+          if (this.indices(k1) != v.indices(k2) || this.values(k1) != v.values(k2)) {
+            return false
+          }
+          k1 += 1
+          k2 += 1
+        }
+        throw new Exception("unreachable")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: I wondered to myself whether this could be simplified to not have `while (true)`, the dummy `Exception`, etc. The best I could do was with a helper function:

```
...
var k1 = nextNonzero(this.values, 0)
var k2 = nextNonzero(v.values, 0)

while (k1 < this.values.size && k2 < v.values.size) {
  if (this.indices(k1) != v.indices(k2) || this.values(k1) != v.values(k2)) {
    return false
  }
  k1 = nextNonzero(this.values, k1 + 1)
  k2 = nextNonzero(v.values, k2 + 1)
}

return (k1 == this.values.size && k2 == v.values.size) 
...

def nextNonzero(values: Array[Double], from: Int): Int = {
  var index = from
  while (index < this.values.size && this.values(index) == 0.0) index += 1
  index
}
```

I'm not sure it's better, just food for thought.

So the idea would be to specialize `hashCode` as well, and also handle `DenseVector` right? and even remove the implementations in the parent?

**Comment by hhbyyh** [comment]
@srowen Thanks a lot for the improvement. The code do look cleaner and the helper function `nextNonzero` can also be used in the `hashCode` function. Great idea.

I'll use it if you don't mind. And `this` in the `nextNonzero` can be removed, right?

**Comment by srowen** [comment]
@hhbyyh Oops yes that's a typo. The helper function should refer to the local argument `values` only!

**Comment by mengxr** [comment]
@hhbyyh @srowen There are some performance issues if we use unnecessary index lookup. Having many `other.values(i)` calls is slower than `val otherValues = other.values` and then many `otherValues(i)` calls. I'm suggesting some code like the following (I didn't try compiling the code):

~~~scala
var ii0 = this.indices
var vv0 = this.values
var ii1 = this.indices
var vv1 = this.values
var j0 = 0
var j1 = 0
var i0 = 0
var i1 = 0
var v0 = 0.0
var v1 = 0.0
var pj0 = -1
var pj1 = -1
var allEqual = true
while(allEqual && j0 < sz0 && j1 < sz1) {
  if (pj0 < j0) {
    i0 = ii0(j0)
    v0 = vv0(j0)
    pj0 = j0
  }
  if (pj1 < j1) {
    i1 = ii1(j1)
    v1 = vv1(j1)
    pj1 = j1
  }
  if (i0 == i1) {   
    allEqual &&= v0 == v1
    j0 += 1
    j1 += 1
  } else if (i0 < i1) {
    allEqual &&= v0 == 0.0
    j0 += 1
  } else {
    allEqual &&= v1 == 0.0
    j1 += 1
  }
while (allEqual & j0 < sz0) {
  allEqual &&= vv0(j0) == 0.0
  j0 += 1
}
while (allEqual & j1 < sz1) {
  allEqual &&= vv1(j1) == 0.0
  j1 += 1
}
allEqual
~~~

**Comment by hhbyyh** [comment]
Oh I didn't see your comment before the PR. @mengxr I surely found a lot of the code using the pattern (first assign locally, then access) in Spark, and we can follow it. 

Just to be honest, I'm not sure I completely understand the reason, is it for sparing a memory addressing? I ran some local perf test and got no obvious difference.  Would you please share some insight? Thanks a lot!

**Comment by hhbyyh** [comment]
@mengxr I did some further test and find the perf difference caused by the heap access. Thanks for pointing it out.

Sent an update containing the locality improvement. @srowen  you may notice I changed the implementation back a little and uses the allEqual inspired by @mengxr :)

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 4]
```diff
@@ -57,7 +57,19 @@ sealed trait Vector extends Serializable {
     }
   }
 
-  override def hashCode(): Int = util.Arrays.hashCode(this.toArray)
+  override def hashCode(): Int = {
+    var result: Int = size + 31
+    this.foreachActive{case(index, value) =>
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: You should put spaces around the `{` and in `if(...` below. I think you could note where this hash function comes from (the JDK right?) to explain it. But yes this is looking right.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 5]
```diff
@@ -449,6 +461,37 @@ class SparseVector(
...
+          allEqual = thisIndices(k1) == otherIndices(k2) && thisValues(k1) == otherValues(k2)
+          k1 += 1
+          k2 += 1
+        }
+        allEqual
+      }
+      case _ => super.equals(other)
+    }
+  }
+
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Yes I like this version. Still pretty clear, and I think it's fine to optimize this a bit. There's still an opportunity to optimize sparse vs dense comparison, right?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala (hunk 4 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Noted. Yes it's from JDK. I'll add comment.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala (hunk 5 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Yes, about the sparse vs dense.
My current thought is put a new function in `Vectors`, maybe something like `sqdist(sparse, dense)`, and call it from `Vector.equals`, does it sound good?


Yes, about the sparse vs dense.
My current thought is put a new function in `Vectors`, maybe something like `sqdist(sparse, dense)`, and call it from `Vector.equals`, does it sound good?

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: You mean a function structured like `sqdist`, not a distance-based function right? yes that sounds good to me.

Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: yes, and I found the sparse vs dense is actually quite similar to sparse vs sparse, I'm trying to unify them.

**Comment by hhbyyh** [comment]
Sent an update unifying the "sparse vs sparse" and "sparse vs dense".

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 6]
```diff
@@ -51,13 +51,35 @@ sealed trait Vector extends Serializable {
...
+      case v2: Vector => {
+        if (this.size != v2.size) return false
+        (this, v2) match {
+          case (s1: SparseVector, s2: SparseVector) =>
+            Vectors.getEquality(s1.indices, s1.values, s2.indices, s2.values)
+          case (s1: SparseVector, d1: DenseVector) =>
+            Vectors.getEquality(s1.indices, s1.values, 0 until d1.size, d1.values)
+          case (d1: DenseVector, s1: SparseVector) =>
+            Vectors.getEquality(0 until d1.size, d1.values, s1.indices, s1.values)
+          case (_, _) => util.Arrays.equals(this.toArray, v2.toArray)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Yeah looking good. I have only minor things to say, which don't necessarily need to change. For example the utility method could be called `Vectors.equals`. Should there even be a catch-all case? I wonder if failing fast is good, if no other implementations are expected now.

File: mllib/src/test/scala/org/apache/spark/mllib/linalg/VectorsSuite.scala [hunk 7]
```diff
@@ -89,6 +89,29 @@ class VectorsSuite extends FunSuite {
...
+    }
+
+    val another = Vectors.sparse(10, Array(1, 3, 4, 7), Array(0.9, 0.2, 0.7, 0.6))
+
+    for (v <- vectors) {
+      assert(v != another)
+      assert(v.## != another.##)
+    }
+  }
+
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/mllib/linalg/VectorsSuite.scala]: This still needs tests involving dense vectors right?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala (hunk 6 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Thanks for review, @srowen 
Yes, I can change the name. Just like the name in `util.Arrays.equals`.
And for the catch-all case, I was thinking `util.Arrays.equals` is also fail-fast and it can cover the dense vs dense case well, and with some extensiblity. Maybe I don't catch your point. Let me know if still we should change this. Thanks

File: mllib/src/test/scala/org/apache/spark/mllib/linalg/VectorsSuite.scala (hunk 7 above)
Comment (hhbyyh) [inline mllib/src/test/scala/org/apache/spark/mllib/linalg/VectorsSuite.scala]: sure. I'll add some.

**Comment by hhbyyh** [comment]
New updated added for ut update and function name change.

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Comment by hhbyyh** [comment]
Thanks
